builtin/credential/cert/path_login.go

```diff
@@ -147,19 +154,28 @@ func (b *backend) verifyCredentials(req *logical.Request) (*ParsedCert, *logical
 		return nil, logical.ErrorResponse("client certificate must be supplied"), nil
 	}
 
+	// Allow constraining the login request to a single CertEntry
+	certName := d.Get("name").(string)
+
 	// Load the trusted certificates
-	roots, trusted, trustedNonCAs := b.loadTrustedCerts(req.Storage)
+	roots, trusted, trustedNonCAs := b.loadTrustedCerts(req.Storage, certName)
 
 	// If trustedNonCAs is not empty it means that client had registered a non-CA cert
 	// with the backend.
 	if len(trustedNonCAs) != 0 {
-		policy := b.matchNonCAPolicy(connState.PeerCertificates[0], trustedNonCAs)
-		if policy != nil && !b.checkForChainInCRLs(policy.Certificates) {
-			return policy, nil, nil
+		clientCert := connState.PeerCertificates[0]
+		for _, trustedNonCA := range trustedNonCAs {
+			tCert := trustedNonCA.Certificates[0]
+			// Check for client cert being explicitly listed in the config (and matching other constraints)
+			if tCert.SerialNumber.Cmp(clientCert.SerialNumber) == 0 &&
+				bytes.Equal(tCert.AuthorityKeyId, clientCert.AuthorityKeyId) &&
+				b.matchesConstraints(connState.PeerCertificates[0], trustedNonCA.Certificates, trustedNonCA) {
```

Comment:
I was just matching the previous rule (check against CRLs), and it didn't seem wrong to also check Name constraints (even if it is an uncommon use case).